Disable AllowSynchronousIO by default in all servers

Tratcher opened the pull request:
#4774 #5874
I had to bump IIS's dependencies to get the response compression fix. This will go away when I can rebase on the refs changes.

Let's see how many upstream tests fail...

Blocked by MVC #6397.

Tratcher commented:
Rebase on https://github.com/aspnet/AspNetCore/pull/4962

Tratcher commented:
https://github.com/aspnet/AspNetCore/blob/master/src/Middleware/CORS/samples/SampleDestination/SampleMiddleware.cs

Tratcher commented:
Ready for review. The failures are known logging issues @pakrym is dealing with.

Tratcher commented:
Hmm, this should have caused more failures, our test code isn't that clean... 

I'm going to add the same feature to the test server and see what happens.

jkotalik commented on build/dependencies.folderbuilds.props:
```diff
@@ -1,4 +1,4 @@
-﻿<!-- This file is temporary while we rework dependencies management. It is used when building individual folders as if they were repos. -->
+<!-- This file is temporary while we rework dependencies management. It is used when building individual folders as if they were repos. -->
```
nit remove diff.

jkotalik commented on src/Servers/Kestrel/test/InMemory.FunctionalTests/RequestTests.cs:
```diff
@@ -1000,14 +1000,14 @@ public async Task DoesNotEnforceRequestBodyMinimumDataRateOnUpgradedRequest()
 
         [Fact]
-        public async Task SynchronousReadsAllowedByDefault()
+        public async Task SynchronousReadsDisallowedByDefault()
```
This test seems to be doing way more than described.  Why is there a difference between first and second requests? Can we consider adding a simpler test that just checks:

```c#
var ioEx = Assert.Throws<InvalidOperationException>(() => context.Request.Body.Read(new byte[1], 0, 1));
Assert.Equal(CoreStrings.SynchronousReadsDisallowed, ioEx.Message);
```

And change the name of the test to what it is actually testing?

Tratcher replied:
The second request opts into sync reads via bodyControlFeature.

jkotalik replied:
Does it need the first request to begin with? Can the first request opt just do a sync read on the body?

Tratcher replied:
I've split this into two tests.

jkotalik commented on src/Servers/IIS/test/testassets/InProcessWebSite/Startup.cs:
```diff
@@ -347,10 +347,10 @@ private async Task ReadAndWriteEchoLines(HttpContext ctx)
 
             var reader = new StreamReader(ctx.Request.Body);
-            while (!reader.EndOfStream)
+            while (true)
```
Why was this change made?

Tratcher replied:
EndOfStream calls sync Read and throws.

halter73 commented on src/Servers/IIS/test/IIS.Tests/HttpBodyControlFeatureTests.cs:
```diff
@@ -10,22 +11,43 @@ namespace Microsoft.AspNetCore.Server.IISIntegration.FunctionalTests
+            Exception readException = null;
+            using (var testServer = await TestServer.Create(
+                ctx => {
+                    var bodyControl = ctx.Features.Get<IHttpBodyControlFeature>();
```
Nit: this variable isn't used

halter73 commented on src/Servers/IIS/test/testassets/InProcessWebSite/Startup.cs:
```diff
@@ -347,10 +347,10 @@ private async Task ReadAndWriteEchoLines(HttpContext ctx)
             {
                 var line = await reader.ReadLineAsync();
-                if (line == "")
+                if (line == "" || line == null)
```
Nit: Any empty string just indicates there was an empty line, not the end of the request body. I think it's correct to only check for null if you're trying to echo the body.

konraddysput replied:
Maybe its silly question - why you didnt use `string.IsNullOrEmpty` ?

halter73 commented on src/Servers/IIS/test/testassets/InProcessWebSite/Startup.cs:
```diff
@@ -454,6 +454,7 @@ private async Task UpgradeFeatureDetection(HttpContext ctx)
         private async Task TestReadOffsetWorks(HttpContext ctx)
         {
             var buffer = new byte[11];
+            ctx.Features.Get<IHttpBodyControlFeature>().AllowSynchronousIO = true;
```
Might as well just use ReadAsync in this test

Tratcher commented on src/Middleware/CORS/samples/SampleDestination/Startup.cs:
```diff
@@ -86,9 +86,7 @@ private Task HandleRequest(HttpContext context)
-            context.Response.Body.Write(content, 0, content.Length);
-
-            return Task.CompletedTask;
+            return context.Response.Body.WriteAsync(content, 0, content.Length);
```
? I haven't seen any consistent pattern for that.

danielmarbach commented on src/Servers/IIS/src/Microsoft.AspNetCore.Server.IIS/IISServerOptions.cs:
```diff
@@ -1,10 +1,20 @@
 // Copyright (c) .NET Foundation. All rights reserved.
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
+using Microsoft.AspNetCore.Http;
```
Redundant using?

Tratcher replied:
No, that's there fro the doc comment cref.

NinoFloris commented:
What's left of the ambition to expose the pipes instead of sticking with streams?

Tratcher commented:
@NinoFloris that's still happening. https://github.com/aspnet/AspNetCore/issues/4757

Tratcher commented:
```
2018-12-20T00:53:00.2990733Z   Failed   Microsoft.AspNetCore.Server.IISIntegration.FunctionalTests.CompressionModuleTests.DynamicResponsesAreCompressed
2018-12-20T00:53:00.2991617Z   Error Message:
2018-12-20T00:53:00.3000411Z    Assert.Equal() Failure
2018-12-20T00:53:00.3007589Z                                      (pos 20)
2018-12-20T00:53:00.3007782Z   Expected: Message1\r\nMessage2\r\n
2018-12-20T00:53:00.3012128Z   Actual:   Message1\r\nMessage2\r\n\r\n
2018-12-20T00:53:00.3015538Z                                      (pos 20)
```
All other failures look like known unrelated issues.

davidfowl commented:
> What's left of the ambition to expose the pipes instead of sticking with streams.

There are a bunch of work items for doing this in the 3.0 milestone. We think we have a plan that is both backwards compatible and exposes pipes with streams

Tratcher commented:
Blocked by MVC https://github.com/aspnet/AspNetCore/issues/6397. @davidfowl how about I turn the setting back off by default but merge the other improvements I've made so far like the test fixes, the new test host feature, etc.?

davidfowl commented:
Sounds good to me

Tratcher committed:
Disable AllowSynchronousIO by default in all servers #4774

Tratcher commented:
Split. https://github.com/aspnet/AspNetCore/pull/6404 deals with the improvements so far and this PR only deals with changing the defaults.

davidfowl commented:
The other thing we can do is turn it off in those code paths to keep everything clean.

Tratcher commented:
I doubt that would be clean. Hundreds of MVC tests are affected on many different code paths.

davidfowl commented:
I mean no new broken logic would sneak in if we can turn it on globally and enable sync IO on those code paths.

Tratcher committed:
Fix tests

Tratcher commented:
Ready for a final review. I've added some fixes and workarounds in MVC, and cleaned up a few more tests elsewhere.

(The Helix failures are unrelated)

halter73 reviewed (approved):
Double approved.
